Wangscape/tilegen/CartesianPower.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+#include "CartesianPowerIterator.h"
+#include <utility>
+
+namespace tilegen
+{
+
+template<typename InputIt>
+class CartesianPower
+{
+public:
+    static_assert(std::is_base_of<std::input_iterator_tag, 
+                                  typename std::iterator_traits<InputIt>::iterator_category>::value,
+                  "InputIt must be at least an InputIterator");
```

Comment:
The `first` iterator is used in code equivalent to the following:
```
x = first;
*x;
++x;
*x;
x = first; // *
*x;
```
If `first` is an `InputIterator`, `first` may be invalidated before line `*` is called.

The template parameter for `CartesianPower` and `CartesianPowerIterator` should be renamed `FowardIt` and `static_assert`s should check for `std::forward_iterator_tag`, which supports multi-pass iteration.